### PR TITLE
api: Return custom metadata in stat_object.

### PR DESCRIPTION
**Pull request by harshavardhana**
We allow custom metadata to be set
in fput_object(). So we shall properly
return custom metadata back as requested
by the caller.

Fixes https://github.com/minio/minio/issues/4136

### CODE REVIEW

File: minio/api.py [hunk 1]
```diff
@@ -1045,14 +1045,35 @@ def stat_object(self, bucket_name, object_name):
...
+            'cache-control',
+            'content-encoding',
+            'content-disposition',
+            ## Add more supported headers here.
+        ]
+
+        ## Capture only custom metadata.
+        custom_metadata = dict()
+        for k in response.headers:
+            if k in supported_headers:
```
Comment (krisis) [inline minio/api.py]: We can combine the two conditions as,
```python
if k in supported_headers or k.lower().startswith('x-amz-meta'):
    custom_metadata[k] = response.headers.get(k)
```

Comment (harshavardhana) [inline minio/api.py]: sure @krisis good idea.

**Comment by harshavardhana** [comment]
PR is updated.

File: minio/api.py [hunk 2]
```diff
@@ -1045,14 +1045,34 @@ def stat_object(self, bucket_name, object_name):
...
+            'cache-control',
+            'content-encoding',
+            'content-disposition',
+            ## Add more supported headers here.
+        ]
+
+        ## Capture only custom metadata.
+        custom_metadata = dict()
+        for k in response.headers:
+            if (k in supported_headers or
```
Comment (krisis) [inline minio/api.py]: what is the need to paranthesise the expression in if-block?

Comment (harshavardhana) [inline minio/api.py]: Indentation

Comment (harshavardhana) [inline minio/api.py]: Remove the indentation block.

File: tests/functional/tests.py [hunk 3]
```diff
@@ -133,6 +133,13 @@ def main():
     # Get a full object locally.
     client.fget_object(bucket_name, object_name, 'newfile-f')
 
+    client.fput_object(bucket_name, object_name+'-f', 'testfile',
+                       metadata={'x-amz-meta-testing': 'value'})
+
+    stat = client.fget_object(bucket_name, object_name+'-f', 'newfile-f-custom')
+    if not stat.metadata.has_key('X-Amz-Meta-Testing'):
```
Comment (vadmeste) [inline tests/functional/tests.py]: Can we also check the value of this header ?

File: docs/API.md [hunk 4]
```diff
@@ -645,6 +645,7 @@ __Return Value__
 |``obj.etag``|_string_|etag of the object.|
 |``obj.content_type``|_string_  | Content-Type of the object.|
 |``obj.last_modified``|_time.time_  |modified time stamp.|
+|``obj.metadata`` |_dict_ | Contains any additional metadata on the object. |
```
Comment (vadmeste) [inline docs/API.md]: Will we bump a new major version ?

File: minio/api.py [hunk 5]
```diff
@@ -1045,14 +1045,33 @@ def stat_object(self, bucket_name, object_name):
...
+            'cache-control',
+            'content-encoding',
+            'content-disposition',
+            ## Add more supported headers here.
+        ]
+
+        ## Capture only custom metadata.
+        custom_metadata = dict()
+        for k in response.headers:
+            if k in supported_headers or k.lower().startswith('x-amz-meta'):
```
Comment (harshavardhana) [inline minio/api.py]: Sure

File: docs/API.md (hunk 4 above)
Comment (harshavardhana) [inline docs/API.md]: No we don't need to bump major version the type doesn't change neither it breaks anything. It is just another field added.

Comment (vadmeste) [inline docs/API.md]: oups yes.

File: tests/functional/tests.py (hunk 3 above)
Comment (harshavardhana) [inline tests/functional/tests.py]: Fixed

File: minio/api.py [hunk 6]
```diff
@@ -1045,14 +1045,33 @@ def stat_object(self, bucket_name, object_name):
                                   object_name=object_name)
 
         etag = response.headers.get('etag', '').replace('"', '')
+        response.headers.pop('etag');
```
Comment (vadmeste) [inline minio/api.py]: One last comment, not sure what are you doing here? removing header ? for what reason?

Comment (harshavardhana) [inline minio/api.py]: Actually the idea was earlier to just save the response.headers verbatim. But since we are filtering now its not needed.

Comment (harshavardhana) [inline minio/api.py]: Fixed good catch @vadmeste

**Pull request merged**